src/user/AddressBookController.ts

```diff
@@ -28,13 +28,15 @@ export interface ContactEntry {
  * @property chainId - Chain id identifies the current chain
  * @property memo - User's note about address
  * @property isEns - is the entry an ENS name
+ * @property addressType - is the type of this address
  */
 export interface AddressBookEntry {
   address: string;
   name: string;
   chainId: string;
   memo: string;
   isEns: boolean;
+  addressType: string;
```

Comment:
Could we use an enum here rather than a string? What are the possible types?

Reply:
Sure, there are two types 
'EXTERNALLY_OWNED_ACCOUNTS' and 'CONTRACT_ACCOUNTS'

Reply:
See [this comment](https://github.com/MetaMask/metamask-extension/issues/12683#issuecomment-1144888878), I think this issue has been misunderstood. Distinguishing between accounts and contract accounts would not solve the referenced bug.

Reply:
@Gudahtt Do you still have the same concerns about this PR?

Reply:
Yes. It looks like this PR hasn't been updated to reflect the decision we made on how to move forward with this, because there are still only two types of addresses listed here.

I have no problem with tracking whether we think an address is an externally owned account or a contract account. But that's not what the linked issue called for. For that, we need to distinguish _accounts_ (including contract accounts and externally owned accounts) from non-accounts (e.g. contracts that are not contract accounts).

Having an enum for "unknown" would be useful as well.

Reply:
"EXTERNALLY_OWNED_ACCOUNTS" and "CONTRACT_ACCOUNTS" both can be non-accounts, so we will add a new enum "UNKOWN" and link that enum with non-accounts.
We won't list addresses with addressType "unknown" in the user's AddressBook.

Did I understand correctly?

Reply:
>"EXTERNALLY_OWNED_ACCOUNTS" and "CONTRACT_ACCOUNTS" both can be non-accounts

Could you elaborate on this? I had expected both of these to represent accounts. They both have 'account' in the name.

How would a contract that is not an account be represented?

Reply:
It will be represented as NON_ACCOUNT.

Reply:
EXTERNALLY_OWNED_ACCOUNTS - valid account, it will be listed in the address book
CONTRACT_ACCOUNTS - valid contract account, it will be listed in the address book
NON_ACCOUNTS - invalid accounts that we want to filter out from the address book list (_get_code returns contract_ || _get_code returns EOA and nonce = 0_)
UNKNOWN - default type for accounts that are in the address book before this PR change. Once the user interacts (e.g. click on the account from the address book) with the account that he had before we will update addressType of that account.

Reply:
Ah, it seems I replied too quickly, before your edit.

Those values make sense to me. I'm not sure on the nonce-checking strategy, but we can talk about that in the extension PR/issue.